src/middlewared/middlewared/plugins/smb_/registry_global.py

```diff
@@ -0,0 +1,297 @@
+        ret['ad'] = DEFAULT_AD_PARAMETERS.keys()
+        ret['ldap'] = DEFAULT_LDAP_PARAMETERS.keys()
+        return ret
+
+    @private
+    async def strip_idmap(self, reg_defaults):
+        """
+        All params related to idmap backends will be handled
+        in idmap plugin.
+        """
+        idmap_params = {}
+        for k, v in reg_defaults.items():
+            if k.startswith("idmap config"):
+                idmap_params[k] = v
+
+        for e in idmap_params.keys():
+            reg_defaults.pop(e, "")
+
+        return idmap_params
+
+    @private
+    async def strip_directory_services(self, reg_defaults):
+        def_ds_params = []
+        def_ds_params.extend(DEFAULT_AD_PARAMETERS.keys())
+        def_ds_params.extend(DEFAULT_LDAP_PARAMETERS.keys())
+        ds_params = {}
+
+        for k, v in reg_defaults.items():
+            if k in def_ds_params:
+                ds_params[k] = v
+
+        for e in ds_params.keys():
+            reg_defaults.pop(e, "")
+
+        return ds_params
+
+    @private
+    async def reg_globals(self):
+        """
+        Split smb.conf parameters into portions used by relevant plugins.
+
+        `raw` contains unmodified smb.conf
+        `idmap` contains idmap configuration
+        `ds` contains directory service configuration
+        `smb` contains smb service configuation (smb plugin)
+        """
+        ret = {}
+        """
+        reg_showshare will fail for `global` if registry has no global entries.
+        In this case simply return an empty config (since it's actually empty anyway).
+        """
+        try:
+            global_conf = await self.middleware.call('sharing.smb.reg_showshare', 'global')
+        except Exception:
+            self.logger.debug("Failed to retrieve global share config from registry")
+            return {'raw': {}, 'idmap': {}, 'ds': {}, 'smb': {}}
+
+        ret['raw'] = global_conf.copy()
+        ret['idmap'] = await self.strip_idmap(global_conf)
+        ret['ds'] = await self.strip_directory_services(global_conf)
+        ret['smb'] = global_conf
+        return ret
+
+    @private
+    async def reg_config(self):
+        """
+        This co-routine is called in smb.config() when cluster support is enabled.
+        In a clustered configuration, we rely exclusively on the contents of the
+        clustered SMB configuration in Samba's registry.
+        """
+        reg_globals = (await self.middleware.call('smb.reg_globals'))['smb']
+        bind_ips = (reg_globals.pop("interfaces", "")).split()
+        if bind_ips:
+            bind_ips.remove("127.0.0.1")
+
+        reg_globals.pop("bind interfaces only", "")
+        loglevel = reg_globals.pop("log level", "1")
+        if loglevel.startswith("syslog@"):
+            loglevel = loglevel[len("syslog@")]
+
+        llevel = LOGLEVEL_MAP.get(loglevel.split()[0])
+
+        ret = {
+            "id": 1,
+            "netbiosname": reg_globals.pop("tn:netbiosname"),
```

Comment:
On my box I'm getting multiple `KeyErrors`

```suggestion
            "netbiosname": reg_globals.pop("tn:netbiosname", "truenas"),
```